### PR TITLE
DEVADV-192 n1ql queries with sdk page validation/update

### PR DESCRIPTION
**Pull request by mgroves**
I've read through this document and made some improvements/fixes. Most notably, some of the code samples were incorrect.

I've also checked in the full .net project that this page is pulling code samples from. In the future, when it's possible to point to a complete code sample, this page should be updated.

### CODE REVIEW

File: modules/howtos/pages/n1ql-queries-with-sdk.adoc [hunk 1]
```diff
@@ -11,115 +11,82 @@ Querying can solve typical programming tasks such as finding a user profile by e
...
 Our query service uses N1QL, which will be fairly familiar to anyone who's used any dialect of SQL.
 xref:#additional-resources[Further resources] for learning about N1QL are listed at the bottom of the page.
-Before you get started you may wish to checkout the xref:6.5@server:n1ql:n1ql-language-reference/index.adoc[N1QL intro page], or just dive in with a query against our travel sample data set.
+Before you get started you may wish to checkout the xref:server:n1ql:n1ql-language-reference/index.adoc[N1QL intro page], or just dive in with a query against xref:server:manage:manage-settings/install-sample-buckets.adoc[our travel sample data set].
 In this case, the one thing that you need to know is that in order to make a Bucket queryable, it must have at least one index defined.
-You can define a _primary_ index on a bucket.
-When a primary index is defined you can issue non-covered queries on the bucket as well.
+You can define a _primary_ index on a bucket. When executing queries, if a suitable index is not found, the primary index will ensure that the query will be executed anyway (the primary index should be not be used in production to prevent scanning of the whole bucket).
```
Comment (AV25242) [inline modules/howtos/pages/n1ql-queries-with-sdk.adoc:17]: the primary index should **be** not be used.... please remove the extra **be**

File: modules/howtos/pages/n1ql-queries-with-sdk.adoc [hunk 2]
```diff
@@ -11,115 +11,82 @@ Querying can solve typical programming tasks such as finding a user profile by e
...
 
-or replace _travel-sample_ with a different Bucket name to build an index on a different dataset.
+or replace _travel-sample_ with a different Bucket name to build an index on a different Bucket.
 
-NOTE: The default installation places cbq in `/opt/couchbase/bin/` on Linux, `/Applications/Couchbase Server.app/Contents/Resources/couchbase-core/bin/cbq` on OS X, and `C:\Program Files\Couchbase\Server\bin\cbq.exe` on Microsoft Windows.
+NOTE: If you are using _travel-sample_, it comes with a primary index already created.
 
-// TODO: improving this and/or adding server/6.0/tools/query-workbench.html
 
+== Parameterization
```
Comment (AV25242) [inline modules/howtos/pages/n1ql-queries-with-sdk.adoc]: Queries & Placeholders is currently being used across all SDK except Java, I agree that this needs to be changed. Can you make it consistent  and call this section **Paramerterized Queries**

File: modules/howtos/pages/n1ql-queries-with-sdk.adoc [hunk 3]
```diff
@@ -11,115 +11,82 @@ Querying can solve typical programming tasks such as finding a user profile by e
...
 
-== Queries & Placeholders
-
-Placeholders allow you to specify variable constraints for an otherwise constant query.
-There are two variants of placeholders: postional and named parameters.
-Positional parameters use an ordinal placeholder for substitution and named parameters use variables.
-A named or positional parameter is a placeholder for a value in the WHERE, LIMIT or OFFSET clause of a query.
-Note that both parameters and options are optional.
+Parameters allow you to specify variable constraints for an otherwise constant query. Parameterization is also a good defense against SQL injection.
+There are two variants of parameters: positional and named.
```
Comment (AV25242) [inline modules/howtos/pages/n1ql-queries-with-sdk.adoc:34]: good point on SQL Injection, will help to provide a [link](https://owasp.org/www-community/attacks/SQL_Injection)

File: modules/howtos/examples/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                // error
+            }
+
+            // iterate over rows
+            // NOTE: because the query is using AtPlus, the new user will be indexed, but notice
+            // the extra step of creating the 'state' object and passing it in as a 'ConsistentWith' option
+            await foreach (var row in result)
+            {
+                // each row is an instance of the Query<T> call (e.g. dynamic or custom type)
+                var name = row.name;        // "Mike"
```
Comment (AV25242) [inline modules/howtos/examples/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs]: should this not be Mike AtPlus ?

File: modules/howtos/examples/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            {
+                // error
+            }
+
+            // iterate over rows
+            // NOTE: because the query is using RequestPlus, the new user will be indexed, but the query may take slightly longer to finish
+            await foreach (var row in result)
+            {
+                // each row is an instance of the Query<T> call (e.g. dynamic or custom type)
+                var name = row.name;        // "Mike"
```
Comment (AV25242) [inline modules/howtos/examples/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs]: Should the output not be mike RequestPlus as this is remarks consistency ?

File: modules/howtos/pages/n1ql-queries-with-sdk.adoc [hunk 6]
```diff
@@ -11,115 +11,82 @@ Querying can solve typical programming tasks such as finding a user profile by e
...
-}
+include::example$/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs[tag=namedparameter]
+
+include::example$/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs[tag=namedparameter2]
 ----
 
 == Scan Consistency
 
-
 Setting a staleness parameter for queries, with `scan_consistency`, enables a tradeoff between latency and (eventual) consistency.
```
Comment (AV25242) [inline modules/howtos/pages/n1ql-queries-with-sdk.adoc:122]: Great examples, however reading it as a new developer have the following questions 
a) Are we starting of with an assumption that a document with user name "Mike" already exist ?
b) I see that the code examples has comments but can that be written against each consistency level.

File: modules/howtos/examples/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            {
+                // error
+            }
+
+            // iterate over rows
+            // NOTE: because the query is using NotBounded, the new user may not be indexed yet
+            await foreach (var row in result)
+            {
+                // each row is an instance of the Query<T> call (e.g. dynamic or custom type)
+                var name = row.name;        // "Mike"
```
Comment (AV25242) [inline modules/howtos/examples/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs]: Is Mike in a document that already exist ? If thats the case can we have a mention of this somewhere ?

File: modules/howtos/pages/n1ql-queries-with-sdk.adoc [hunk 8]
```diff
@@ -11,115 +11,82 @@ Querying can solve typical programming tasks such as finding a user profile by e
...
-var state = MutationState.From(upsertResult);
-
-// use mutation state with query option
-var result = await cluster.QueryAsync<dynamic>(
-    "SELECT x.* FROM `default` WHERE x.Type=$1",
-    options => options.ConsistentWith(state)
-                      .Parameter("User")
-);
+include::example$/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs[tag=atplus]
 ----
```
Comment (AV25242) [inline modules/howtos/pages/n1ql-queries-with-sdk.adoc:148]: We should add a section that tells us the different QueryOptions that are available today, just like the Java document does
https://docs.couchbase.com/java-sdk/3.0/howtos/n1ql-queries-with-sdk.html

File: modules/howtos/examples/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        {
+            // tag::positional[]
+            var result = await cluster.QueryAsync<dynamic>(
+                "SELECT t.* FROM `travel-sample` t WHERE t.type=$1",
+                options => options.Parameter("landmark")
+            );
+            // end::positional[]
+
+            // check query was successful
+            if (result.MetaData.Status != QueryStatus.Success)
```
Comment (AV25242) [inline modules/howtos/examples/n1ql-queries-sdk/n1ql-queries-sdk/Program.cs:226]: Can we also add a section about what to expect out of a query result ?  There is a Handling results section that can probably changed to Query Results with different options available to us !

**Review by AV25242 — CHANGES_REQUESTED** [review]
I carefully went through the changes as someone without prior CB experience, overall very nice effort. I have few comments that will help immensely and will make it consistent with Java docs.

**Comment by jeffrymorris** [comment]
@mgroves  - I also have an examples project locally with examples for "Connection Management" page which I am _still working on_ (yes @RichardSmedley its coming!). Once this is merged I will pull it and locally merge my stuff into this project.

Thanks,

Jeff

**Comment by mgroves** [comment]
Thanks for the suggestions @av25242 I will work on these next week

File: modules/howtos/pages/n1ql-queries-with-sdk.adoc (hunk 2 above)
Comment (mgroves) [reply modules/howtos/pages/n1ql-queries-with-sdk.adoc]: "Parameterized" but yes that's good :)

File: modules/howtos/pages/n1ql-queries-with-sdk.adoc (hunk 6 above)
Comment (mgroves) [reply modules/howtos/pages/n1ql-queries-with-sdk.adoc:122]: a) Are we starting of with an assumption that a document with user name "Mike" already exist ?

The code samples use `upsert`, so it doesn't assume that. The only time this really matter is for the `NotBounded` example, where running it the first time would (probably) not return a result.

b) I see that the code examples has comments but can that be written against each consistency level.

What kind of comment would you like to see?

**Comment by mgroves** [comment]
I added two big sections per @AV25242's feedback:  Query Results and Query Options, which largely mirror the Java documentation (except translated to the vastly superior :trollface: conventions and idioms of .NET and C#)

File: modules/howtos/pages/n1ql-queries-with-sdk.adoc (hunk 6 above)
Comment (AV25242) [reply modules/howtos/pages/n1ql-queries-with-sdk.adoc:122]: Thanks Matt all changes looks good 

What kind of comment would you like to see?
All i was trying to say is explicitly mention what would be the output in each section of consistency.
you have it but its all in code. But I think thats fine lets move on :)

**Review by AV25242 — APPROVED** [review]

**Review by RichardSmedley — APPROVED** [review]

**Comment by RichardSmedley** [comment]
Thanks @mgroves for all of the improvements - looks good.
I tweaked a couple of links post hoc btw,

**Pull request merged**